components/tikv_alloc/src/trace.rs

```diff
@@ -0,0 +1,220 @@
+    Arc,
+};
+
+type HashMap<K, V> =
+    std::collections::HashMap<K, V, std::hash::BuildHasherDefault<fxhash::FxHasher>>;
+
+#[derive(Debug, PartialEq, Copy, Clone, Hash, Eq)]
+pub enum Id {
+    Name(&'static str),
+    Number(u64),
+}
+
+impl Id {
+    pub fn readable_name(&self) -> String {
+        match self {
+            Id::Name(s) => {
+                let mut s = s.replace('_', "");
+                s.make_ascii_lowercase();
+                let mut res = String::with_capacity(s.len());
+                for p in s.split('_') {
+                    res.push_str(p);
+                    res.push(' ');
+                }
+                res
+            }
+            Id::Number(n) => n.to_string(),
+        }
+    }
+}
+
+impl From<&'static str> for Id {
+    #[inline]
+    fn from(s: &'static str) -> Self {
+        Id::Name(s)
+    }
+}
+
+impl From<u64> for Id {
+    #[inline]
+    fn from(id: u64) -> Self {
+        Id::Number(id)
+    }
+}
+
+impl From<NonZeroU64> for Id {
+    #[inline]
+    fn from(id: NonZeroU64) -> Self {
+        Id::Number(id.into())
+    }
+}
+
+impl Display for Id {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Id::Number(n) => write!(f, "{}", n),
+            Id::Name(n) => write!(f, "{}", n),
+        }
+    }
+}
+
+#[derive(Clone, Copy)]
+pub enum TraceEvent {
+    Add(usize),
+    Sub(usize),
+    Reset(usize),
+}
+
+pub trait MemoryTrace {
+    fn trace(&self, event: TraceEvent);
+    fn snapshot(&self, parent: &mut MemoryTraceSnapshot);
+    fn sub_trace(&self, id: Id) -> Arc<dyn MemoryTrace + Send + Sync>;
+    fn add_sub_trace(&mut self, id: Id, trace: Arc<dyn MemoryTrace + Send + Sync>);
+    fn sum(&self) -> usize;
+}
+
+pub struct MemoryTraceNode {
+    pub id: Id,
+    trace: AtomicUsize,
+    children: HashMap<Id, Arc<dyn MemoryTrace + Send + Sync>>,
+}
+
+impl MemoryTraceNode {
+    pub fn new(id: impl Into<Id>) -> MemoryTraceNode {
+        MemoryTraceNode {
+            id: id.into(),
+            trace: std::sync::atomic::AtomicUsize::default(),
+            children: HashMap::default(),
+        }
+    }
+}
+
+impl MemoryTrace for MemoryTraceNode {
+    fn trace(&self, event: TraceEvent) {
+        match event {
+            TraceEvent::Add(val) => {
+                self.trace.fetch_add(val, Ordering::Relaxed);
+            }
+            TraceEvent::Sub(val) => {
+                self.trace.fetch_sub(val, Ordering::Relaxed);
+            }
+            TraceEvent::Reset(val) => {
+                self.trace.swap(val, Ordering::Relaxed);
+            }
+        }
+    }
+
+    fn snapshot(&self, parent: &mut MemoryTraceSnapshot) {
+        let mut snap = MemoryTraceSnapshot {
+            id: self.id,
+            trace: self.trace.load(Ordering::Relaxed),
+            children: vec![],
+        };
+        for child in self.children.values() {
+            child.snapshot(&mut snap);
+        }
+        parent.children.push(snap);
```

Comment:
Is it better to handle the root node specially?

Reply:
Changed the parameter to an `Option`

Reply:
If it's `None`, how to return the snapshot?

Reply:
😬 Just was still sleepy, now I changed it to return a snapshot.